lib/ecto/repo/schema.ex

```diff
@@ -116,6 +116,19 @@ defmodule Ecto.Repo.Schema do
     end
 
     header = case query.select do
+      %Ecto.Query.SelectExpr{expr: {:%{}, [], [{:|, _, [{:&, _, [ix]}, args]}]}, fields: fields} ->
+        select_fields = for {{:., _, [{:&, _, [^ix]}, field]}, [], []} <- fields, do: field
+        update_fields = Keyword.keys(args)
+
+        select_fields ++ update_fields
+        |> Enum.map(fn field ->
+          case dumper do
+            %{^field => {source, _, false}} -> source
+            %{} -> raise ArgumentError, "cannot select unwritable field `#{field}` for insert_all"
+            nil -> field
+          end
```

Comment:
Should we extract this `case` into a `dump_field` or `dump_insert_field` function?